### DISCUSSION TITLE
How to reset `GetCompilationAsync` cache?

### DISCUSSION THREAD

**Discussion by joshua-light** [dotnet/roslyn#55272] [Q&A] [answered]
Hi! I'm relatively new to the internals of the Roslyn API, trying to build a simple incremental code-generator for a project.

Its idea is to compile the project, get its underlying types and use them to produce new code. Here is how it looks:
```cs
var path = "SomeProject.csproj";
var workspace = MSBuildWorkspace.Create();
var project = await workspace.OpenProjectAsync(path);
var compilation = await project.GetCompilationAsync();

// Do some stuff with `compilation`.
// ...
```

However, the problem is that preparation steps like `OpenProjectAsync` and `GetCompilationAsync` take like 4-5 seconds on a relatively simple project.

The question is: is there any way to open a project once and only recompile it? Or maybe another way to speed up things a bit? `GetCompilationAsync` is meant to return a cached compilation, hence, the question.

**Comment by jasonmalinowski** [comment, marked as answer]
We've been building Source Generators as an API to enable scenarios like this, which both avoids MSBuild problems and also helps performance. Have you looked at those?

**Reply by joshua-light** [reply]
Yes, I'm using Source Generators whenever possible and they're really great. Thanks for this excellent feature!

However, as for now, when working with, for example, Unity, I still need to use an old-fashioned way of doing code-gen. 4-5 seconds for code-gen is not a problem, though, just wanted to be sure I'm not missing some caveat. Thanks!

**Reply by jasonmalinowski** [reply]
So in some sense the performance problem you're experiencing is why we built generators; the GetCompilationAsync() call there still needs to re-read all the source flies and reparse them to get them into your process, and there's no short way around that. If you had a breakdown for whether it's the OpenProjectAsync or GetCompilationAsync that's slower, we might be able to do something with that if we discover OpenProjectAsync is doing more work than it should be.
